File: examease/src/main/java/project/practice/examease/service/QuizService.java
package project.practice.examease.service;

import org.springframework.http.ResponseEntity;
import project.practice.examease.dto.PortalStatDto;
import project.practice.examease.dto.TestDto;
import project.practice.examease.entity.Quiz;

import java.util.List;

public interface QuizService {
    ResponseEntity<String> addQuiz(Quiz requestBody);

    ResponseEntity<Quiz> getQuizById(Integer id);

    ResponseEntity<List<Quiz>> getQuizBySubcategoryId(Integer id);

    ResponseEntity<List<Quiz>> getQuizes();

    ResponseEntity<String> addTest(TestDto requestBody);

    ResponseEntity<PortalStatDto> getPortalStats();


    ResponseEntity<List<Quiz>> getQuizesByPagination(int pageSize, int pageNumber,Integer subcategoryId);

    ResponseEntity<Integer> getQuizCountBySubcategoryId(Integer id);
}

File: exam-ease/src/app/service/quiz.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Quiz, Test } from '../model';

@Injectable({
  providedIn: 'root',
})
export class QuizService {
  // http://localhost:8080/api/examease/subcategories
  constructor(private http: HttpClient) {}
  baseURL: string = 'http://localhost:8080/api/examease';
  getCategories() {
    return this.http.get(this.baseURL + '/categories');
  }
  addQuiz(data: Quiz) {
    return this.http.post(this.baseURL + '/quizes', data);
  }
  getSubcategories() {
    return this.http.get(this.baseURL + '/subcategories');
  }

  getQuizes() {
    return this.http.get(this.baseURL + '/quizes');
  }
  getQuizesByPagination(
    pageSize: Number,
    pageNumber: Number,
    subcategoryId?: Number
  ) {
    if (subcategoryId != undefined || subcategoryId != null) {
      return this.http.get(
        this.baseURL +
          `/quizes-pagination?pageNumber=${pageNumber}&pageSize=${pageSize}&subcategoryId=${subcategoryId}`
      );
    }
    return this.http.get(
      this.baseURL +
        `/quizes-pagination?pageNumber=${pageNumber}&pageSize=${pageSize}`
    );
  }
  getQuizById(id: Number) {
    return this.http.get(this.baseURL + `/quizes/${id}`);
  }
  addTest(data: Test) {
    return this.http.post(this.baseURL + `/tests`, data);
  }
  getPortalStats() {
    return this.http.get(this.baseURL + '/portal-stats');
  }
  getQuizCountBySubcategoryId(subId:Number){
    return this.http.get(this.baseURL+`/quizes/subcategory/count/${subId}`)
  }
}
